Try.onEach

nkoder opened the pull request:
What I found missing in `Try` is possibility to do something on value and return it. For example:
```
Try {
   // ...
}.onEach {
   printLog(it)
}.map {
   // ...
}
```
instead of
```
Try {
   // ...
}.map {
   printLog(it)
   it
}.map {
   // ...
}
```

Commit:
Try.onEach

jpallari commented:
What's the expected behaviour for `onEach` when the given function throws an exception? Example:

```kotlin
Try {
   ...
}.onEach {
  somethingThatThrowsException()
}.map {
  ...
}
```

nkoder commented:
I think it should be wrapped into `Failure` in similar way as it is in `map`. This is the way I implemented it. For example:
```
        val x: Try<String> = success.map {
            throw RuntimeException("here I am")
        }
        x.get() // exception is thrown here
```
This way we preserve the whole reason for using `Try` from funKTionale (or at least the reason I use it), which is to pass exceptions further instead of throwing them in-place.

BTW: another name for `onEach` I saw for this kind of method in other libraries was `peek`

MarioAriasC commented:
I think that `onEach` is on the Kotlin standard library, but `peek` doesn't sound bad.

Can you write a test using the exception scenario?

nkoder commented:
> I think that onEach is on the Kotlin standard library

right :)

> Can you write a test using the exception scenario?

Here it is: https://github.com/MarioAriasC/funKTionale/pull/25/commits/0a93a05de34bf8cea2c141326ed99701c7331698#diff-7ec1a9435e4fb1d41474584cbebf75a7R64
I've written tests in similar way as other tests in same file ( https://github.com/nkoder/funKTionale/blob/0a93a05de34bf8cea2c141326ed99701c7331698/funktionale-try/src/test/kotlin/org/funktionale/tries/TryTest.kt#L59 ) so I used the idea of calling `fail()` inside method. So there is `fail()` inside `onEach { ... }` which is throwing an exception and exception is consumed by `onEach`.

jpallari commented on funktionale-try/src/test/kotlin/org/funktionale/tries/TryTest.kt:
```diff
@@ -59,6 +59,11 @@ class TryTest {
         failure.foreach { fail() }
     }
 
+    @Test fun onEach() {
```
Maybe there should be a test for `success.onEach { throw Exception("expected") }`

jpallari replied:
I suppose that sort of test would also be useful for `forEach` as well. :)

nkoder replied:
Yes, I can rewrite tests for `onEach` (and for `foreach` maybe too) a little bit.

Do you prefer amended commit or a new one? I am a fan of squashing irrelevant changes into one "feature" commit, but maybe you prefer otherwise or it's more convinient to work with separate commits on GitHub (I'm not a GitHub Pull Request veteran ;-) )

MarioAriasC replied:
A new one is ok, whatever you find more convenient. We don't have any standard on that at the moment

jpallari replied:
Sorry for the late response.

Personally, I don't mind either way as long as the commits make sense. I'm not sure what's Mario's opinion on it though.

nkoder replied:
I've pushed one more commit yesterday. Can you look at it?

Commit:
New tests for Try.foreach and Try.onEach: less guessing, more cases. Introduced AssertJ with its nice assertion for throwables.

nkoder commented:
OK, I've managed to find another couple of minutes and rewrote tests for `foreach` and `onEach` in order to: (1) cover more cases, (2) be more explicit without the need for guessing why given test is really testing something.

Now I think it addresses your concerns @MarioAriasC and @jkpl, but... with this approach, it's less aligned with convention I've met in a test file.  I don't know what it's a priority for you.

What is your feedback now? Or maybe it's OK now to merge it to master branch and use it in other apps in a very near future?

jpallari commented on funktionale-try/src/test/kotlin/org/funktionale/tries/TryTest.kt:
```diff
@@ -54,14 +54,75 @@ class TryTest {
+    }
+
+    @Test fun onEach_failure_throwInside() {
+        try {
```
Overall, the tests look good. :+1:

There's one thing to consider: try-catch works sort of like a conditional. If the line `failure.onEach { throw RuntimeException("thrown inside") }.get()` never throws an exception, `assertTrue(e is NumberFormatException)` is not executed. What should be done instead is that the test verifies first that there IS an exception being thrown.

You can leverage [try as an expression](https://kotlinlang.org/docs/reference/exceptions.html#try-is-an-expression) or [funKTionale's Try](https://github.com/MarioAriasC/funKTionale/blob/master/funktionale-try/src/main/kotlin/org/funktionale/tries/Try.kt) for first capturing the exception, and then doing the assertions outside of the `catch` block. There's also [TestNG excepted exceptions](http://www.mkyong.com/unittest/testng-tutorial-2-expected-exception-test/), which might help.

Example format:

```kotlin
val ex: Exception? = try {
  doStuff()
  null
} catch (e: Exception) {
  e
}
assertEquals(ex!!.message, "my message")
```

nkoder replied:
Right, I totally forgot about false positive when nothing is thrown! :) I will push this change as amended commit.

1. "try as an expression": it doesn't look straightforward to me. I mean it will work but will it be easy to understand or more like a clever trick?

2. "funKTionale's Try": sound good although I was never fully convinced to use the same tool for testing and as the subject of a test. Maybe too defensive may way of thinking is :)

3. "TestNG excepted exceptions": sounds really good, but I discovered one caveat (but not a blocker, just a not-a-nice-thing): when thrown exception (`NumberFormatException`) is another than expected, then you get *red* test instead of *orange* (in terms of IntelliJ IDEA coloring). Error message and stacktrace are not about failed assertion, but about the exception thrown inside test (which is what we expect, and we have to figure out that it's about another exception).

4. I will try with another option I love and use everyday: [AssertJ](http://joel-costigliola.github.io/assertj/) I'm aware that it might be unnecessary dependency and extra library to understand, so *feel free* to tell me not to do so ;-) Then I will choose one of options from 1 to 3 :-)

In case of AssertJ `assertThatThrownBy` we have:
* `java.lang.AssertionError: Expecting code to raise a throwable.` if no exception is thrown
* and this in case of wrong exception:
```
java.lang.AssertionError: 
Expecting:
 <java.lang.RuntimeException: thrown inside>
to be an instance of:
 <java.lang.NumberFormatException>
but was instance of:
 <java.lang.RuntimeException>
```

nkoder replied:
Moreover thanks to your comment, @jkpl , I found a false positive in my test for `foreach_failure_throwInside`: it was assertion an exception inside `catch` and `foreach` is not designed to throw an exception. Thanks 👍

nkoder commented:
@jkpl I have amended my second commit with two major changes:

1. Fixed false positive test for `failure.foreach(...)` which was asserting exception in a `catch` block... but `failure.foreach(...)` is not meant to throw an exception.

2. Introduced [AssertJ](http://joel-costigliola.github.io/assertj/index.html)

Sadly, amended commit removed your comment and my broad answer to it :( I thought it worked as on GitLab, where discussion on changes is preserved even if amended commit no longer exists.

I will summarize here shortly what I wrote there:

1. Thanks for your comment about assertions inside `catch` block 👍  I totally forgot about this possibility for false positive tests.

2. All 3 suggested options (try-catch expression, funKTionale Try, TestNG exception testing) seemed to be quite good, although I found some disadvantages in each of them. So I've introduced the 4th solution: AssertJ. This assertions lib has quite descriptive assertions and error messages. And I really love it, and use everyday :-) But... I understand that introducing extra dependency (which introduces another documentation to read) just for some testing corner cases is not quite a best option ;-) So *feel free* to be against it, then I will choose one of your suggested solutions.

jpallari commented:
> Sadly, amended commit removed your comment and my broad answer to it :( I thought it worked as on GitLab, where discussion on changes is preserved even if amended commit no longer exists.

No worries. The comments are still there, but they're hidden. Click on "Show outdated" to view the comments.

> So I've introduced the 4th solution: AssertJ. This assertions lib has quite descriptive assertions and error messages. And I really love it, and use everyday :-) But... I understand that introducing extra dependency (which introduces another documentation to read) just for some testing corner cases is not quite a best option ;-) So feel free to be against it, then I will choose one of your suggested solutions.

AssertJ looks pretty cool! It allows to be more precise on what is being tested compared to the other alternatives. I personally have no problem with the dependency because it only exists in the test build. @MarioAriasC, what do you think?

> "try as an expression": it doesn't look straightforward to me. I mean it will work but will it be easy to understand or more like a clever trick?

I haven't used it much in the context of capturing errors (i.e. `val result: Exception?`), but I've definitely found it to be useful in cases similar to what [the Kotlin documentation demonstrates](https://kotlinlang.org/docs/reference/exceptions.html#try-is-an-expression). I'm slightly biased though because I come from a Scala background, where most things are expressions. :)

MarioAriasC commented on funktionale-try/pom.xml:
```diff
@@ -32,6 +32,12 @@
+            <artifactId>assertj-core</artifactId>
+            <version>3.6.2</version>
+            <scope>test</scope>
+        </dependency>
```
You need to move this to the main pom file. I don't mind new libs if they're just for testing

nkoder replied:
ok

MarioAriasC commented on funktionale-try/src/test/kotlin/org/funktionale/tries/TryTest.kt:
```diff
@@ -54,9 +56,62 @@ class TryTest {
-    @Test fun foreach() {
-        success.foreach { assertEquals(10, it) }
-        failure.foreach { fail() }
+    @Test fun foreach_success_sideEffect() {
```
Not idiomatic naming. You should use:

```kotlin
@Test fun `foreach success side effect`() //or any name that you find convenient and descriptive 
```

nkoder replied:
I was trying to not differ toooo much from original naming. OK, I can change it.

nkoder commented:
Hi @MarioAriasC @jkpl , any news? Any possibility to have this PR merged? :)

MarioAriasC reviewed (changes requested):
Couple of comments

nkoder commented:
@MarioAriasC changed commit is pushed. Please check, if it's OK for you now. And if it is, please tell me when can I expect it to be available to use officially, so I will use it in my project :-)

MarioAriasC commented:
If everything goes well, we could have a 1.0.1 version tonight.

nkoder commented:
BTW: if there is something more to do in funKTionale, then maybe I can sometimes help you guys. Especially if it is kind of clean-up/refactoring stuff. Or maybe it would be better to start updating Wiki to match 1.0 with Try etc.?

MarioAriasC commented:
Wiki work will be greatly appreciated

MarioAriasC commented:
Sorry, I wasn't unable to deploy it. Tomorrow at UK Night time will be

MarioAriasC commented:
Done https://github.com/MarioAriasC/funKTionale/releases/tag/1.0.1

nkoder commented:
thanks :) I've incorporated it in my project already.

About Wiki: where to communicate with you, guys, to not spam here? Or maybe just create pull requests to `funKTionale.wiki` repo?

MarioAriasC commented:
Sorry for the late answer bro, really busy.

It will be really helpful if you create a page on the wiki for ```Try``` 

Cheers
